### PR TITLE
[Seq] Create new clock enabled register op

### PR DESCRIPTION
**Pull request by teqdruid**
Introduces `compreg.ce` which is identical to the existing `compreg` but includes a `clockEnable` signal to enable/disable reading the input. This op makes it easier to create SV with the behavioral syntax rather than the structural mux that we currently construct. Could be lowered into the structural mux if that is desired.

### CODE REVIEW

File: lib/Dialect/Seq/SeqOps.cpp [hunk 1]
```diff
@@ -207,60 +208,75 @@ ParseResult CompRegOp::parse(OpAsmParser &parser, OperationState &result) {
...
-    return parser.resolveOperands(operands, {ty, i1}, loc, result.operands);
-  else
-    return parser.resolveOperands(operands, {ty, i1, i1, ty}, loc,
-                                  result.operands);
+
+  Type i1 = IntegerType::get(result.getContext(), 1);
+  SmallVector<Type, 5> operandTypes;
+  operandTypes.append({ty, i1});
+  if (ClockEnabled)
+    operandTypes.push_back(i1);
```
Comment (mortbopet) [inline lib/Dialect/Seq/SeqOps.cpp:247]: nit: `if constexpr (ClockEnabled)` should apply here

File: lib/Dialect/Seq/SeqOps.cpp [hunk 2]
```diff
@@ -207,60 +208,75 @@ ParseResult CompRegOp::parse(OpAsmParser &parser, OperationState &result) {
       return failure();
   }
 
-  SmallVector<OpAsmParser::UnresolvedOperand, 4> operands;
+  constexpr size_t clockEnabledOffset = (size_t)ClockEnabled;
```
Comment (mortbopet) [inline lib/Dialect/Seq/SeqOps.cpp:211]: nit: I'd prefer `constexpr size_t ceOperandOffset= ClockEnabled ? 1 : 0;` just to be explicit and more readable.

File: lib/Dialect/Seq/SeqOps.cpp [hunk 3]
```diff
@@ -278,6 +294,36 @@ LogicalResult CompRegOp::verify() {
...
+void CompRegClockEnabledOp::getAsmResultNames(OpAsmSetValueNameFn setNameFn) {
+  // If the wire has an optional 'name' attribute, use it.
+  if (!getName().empty())
+    setNameFn(getResult(), getName());
+}
+
+LogicalResult CompRegClockEnabledOp::verify() {
+  if ((getReset() && !getResetValue()) || (!getReset() && getResetValue()))
+    return emitOpError(
+        "either reset and resetValue or neither must be specified");
```
Comment (mortbopet) [inline lib/Dialect/Seq/SeqOps.cpp:314]: `if(getReset() ^ getResetValue())`

File: include/circt/Dialect/Seq/SeqOps.td [hunk 4]
```diff
@@ -51,6 +51,40 @@ def CompRegOp : SeqOp<"compreg",
...
 
+def CompRegClockEnabledOp : SeqOp<"compreg.ce",
+    [Pure, Clocked, Resettable,
+     AllTypesMatch<["input", "data"/*, "resetValue"*/]>,
+     SameVariadicOperandSize,
+     DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]> ]> {
+       // AllTypesMatch doesn't work with Optional types yet.
+
+  let summary = "When enabled, register a value";
+  let description = "See the Seq dialect rationale for a longer description";
```
Comment (mortbopet) [inline include/circt/Dialect/Seq/SeqOps.td:62]: Dialect rationale has not been modified.

File: lib/Dialect/Seq/SeqOps.cpp (hunk 1 above)
Comment (teqdruid) [reply lib/Dialect/Seq/SeqOps.cpp]: Did not know about this!

**Review by mortbopet — APPROVED** [review]

**Pull request merged**